topdown/crypto.go

```diff
@@ -226,6 +271,35 @@ func getX509CertsFromPem(pemBlocks []byte) ([]*x509.Certificate, error) {
 	return x509.ParseCertificates(decodedCerts)
 }
 
+func getRSAPrivateKeyFromString(key string) (*rsa.PrivateKey, error) {
+	// if the input is PEM handle that
+	if strings.HasPrefix(key, "-----BEGIN RSA PRIVATE KEY-----") {
+		return getRSAPrivateKeyFromPEM([]byte(key))
+	}
+
+	// assume input is base64 if not PEM
+	b64, err := base64.StdEncoding.DecodeString(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return getRSAPrivateKeyFromPEM(b64)
+}
+
+func getRSAPrivateKeyFromPEM(pemBlocks []byte) (*rsa.PrivateKey, error) {
+	// decode the pem into the Block struct
+	p, _ := pem.Decode(pemBlocks)
+	if p == nil {
+		return nil, fmt.Errorf("failed to parse PEM block containing the key")
+	}
+
+	if p.Type != blockTypeRSAPrivateKey {
+		return nil, fmt.Errorf("PEM block type is '%s', expected %s", p.Type, blockTypeRSAPrivateKey)
+	}
+
+	return x509.ParsePKCS1PrivateKey(p.Bytes)
```

Comment:
What about PKCS8? We've had issues with that in [the past](https://github.com/open-policy-agent/opa/issues?q=is%3Aissue+pkcs8).

Reply:
According to the go lib `func ParsePKCS8PrivateKey(der []byte) (key interface{}, err error)` is more for all asymmetric keys but not RSA only, and the `func ParsePKCS1PrivateKey(der []byte) (*rsa.PrivateKey, error)` is designed for RSA.

Since this feature is more dedicated for the RSA keys (also function names all prefixed with "RSA"), here the PKCS1 is more suitable, and I think for the PKCS8 we could work in another branch.

Reply:
While you are correct that it's not RSA exclusive, RSA is still valid - and I think a built-in named `parse_rsa_private_key` needs to take this into account. I think the previously linked issues show that. We already have working code to [deal with this](https://github.com/anderseknert/opa/blob/f5d20085d68a0ced7d4dc3eb12764196b450c2b6/internal/jwx/jws/sign/sign.go#L30), so I'd suggest we reuse that. It should not add much in terms of effort, or what do you think?

Reply:
I see what you mean, and I was confused that if we also deal with PKCS8 here that I might have to remove the "RSA" prefix in the function name.

I will add the condition block at the end here which will be similar to what in the `GetSigningKey`

It will be something like this:
```
priv, err := x509.ParsePKCS1PrivateKey(block.Bytes)
		if err != nil {
			pkcs8priv, err2 := x509.ParsePKCS8PrivateKey(block.Bytes)
			if err2 != nil {
				return nil, fmt.Errorf("error parsing private key (%v), (%v)", err, err2)
			}
			return pkcs8priv, nil
		}
		return priv, nil
```